plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -2164,6 +2166,14 @@ private HiveInsertTableHandle beginInsertOrMerge(ConnectorSession session, Conne
                 throw new TrinoException(NOT_SUPPORTED, format("%s Hive table with value of %s property greater than 1 is not supported", description, SKIP_HEADER_COUNT_KEY));
             }
         });
+
+        if (partitionProjectionEnabled) {
+            Optional<PartitionProjection> partitionProjection = getPartitionProjectionFromTable(table, typeManager);
+            if (!partitionProjection.map(PartitionProjection::allowWrite).orElse(true)) {
+                throw new TrinoException(NOT_SUPPORTED, "Hive connector does not support writing such partition with projection");
```

Comment:
Let's see if we can add a more helpful error message here that explains why the operation isn't supported (note how above operations use `description` to indicate whether this is a disallowed insert or merge). Maybe `boolean PartitionProjection#allowWrite()` needs to be more of a `void PartitionProjection#checkWriteSupported(String description)` and throw an exception with a more detailed explanation when rejecting the write.